panel/src/helpers/embed.js

```diff
@@ -0,0 +1,131 @@
+export default {
+  youtube(url) {
+
+    if (!url.match("youtu")) {
+      return false;
+    }
+
+    let uri = null;
+
+    try {
+      uri = new URL(url);
+    } catch (e) {
+      return false;
+    }
+
+    const path = uri.pathname.split("/").filter(item => item !== "");
+    const first = path[0];
+    const second = path[1];
+    const host = "https://" + uri.host + "/embed";
+
+    const isYoutubeId = (id) => {
+      return id.match(/^[a-zA-Z0-9_-]+$/) !== null;
+    };
+
+    let query = uri.searchParams;
+    let src = null;
+
+    switch (path.join("/")) {
+      case "embed/videos":
+      case "playlist":
+        if (isYoutubeId(query.get("list"))) {
+          src = host + "/videoseries";
+        }
+        break;
+      case "watch":
+        if (isYoutubeId(query.get("v"))) {
+          src = host + "/" + query.get("v");
+
+          if (query.has("t")) {
+            query.set("start", query.get("t"));
+          }
+
+          query.delete("v");
+          query.delete("t");
+        }
+
+        break;
+      default:
+        // short URLs
+        if (uri.host.includes("youtu.be")) {
+          src = 'https://www.youtube.com/embed/' + first;
+
+          if (query.has("t")) {
+            query.set("start", query.get("t"));
+          }
+
+          query.delete("t");
+        } else if (first === "embed" && isYoutubeId(second)) {
+          src = host + "/" + second;
+        }
+    }
+
+    if (!src) {
+      return false;
+    }
+
+    const queryString = query.toString();
+
+    if (queryString.length) {
+      src += "?" + queryString.replace("&", "&amp;");
```

Comment:
For cleanliness I think this helper shouldn't escape the string as the returned URL is not a valid URL anymore. Does anything speak against escaping when the URL is used in the component?

Reply:
The only reason for escaping it here is to keep the results identical with the ones in PHP.

Reply:
I just fixed it. I think we should also stop encoding ampersands in PHP. I'm not sure why we are doing it anyway. But this should be done in a different PR. I don't know how deep that problem will go.